[MRG+1] Fix pickling bug  due to multiple inheritance & __getstate__

HolgerPeters opened the pull request:
Includes a reproducing test case.

sklearn.base.BaseEstimator now tries to use other `__getstate__` methods of the class hierarchy first, before defaulting to the `__dict__` attribute

#### Reference Issue
Fix issue #8316

HolgerPeters committed:
Fix pickling bug (issue # 8316) due to multiple inheritance & __getstate__

Includes a reproducing test case.

sklearn.base.BaseEstimator now tries to use other __getstate__ methods of
the class hierarchy first, before defaulting to the __dict__ attribute

lesteve commented on sklearn/base.py:
```diff
@@ -290,10 +290,11 @@ def __repr__(self):
                                                offset=len(class_name),),)
 
     def __getstate__(self):
```
Just wondering whether we should adopt a similar strategy for `__setstate__`? Quickly looking at it, it seems like we are doing `startswith('sklearn.')` there as well to avoid messing with classes outside scikit-learn deriving from `sklearn.base.BaseEstimator`.

jnothman replied:
Yes, I believe there should be parallel changes in `__setstate__`

HolgerPeters replied:
I agree

jnothman commented on sklearn/base.py:
```diff
@@ -290,10 +290,11 @@ def __repr__(self):
                                                offset=len(class_name),),)
 
     def __getstate__(self):
-        if type(self).__module__.startswith('sklearn.'):
```
I think we still want this... we just want to delegate to `super` instead of blindly getting `self.__dict__.items()`.

HolgerPeters replied:
will reintroduce the check

HolgerPeters commented on sklearn/base.py:
```diff
@@ -290,10 +290,15 @@ def __repr__(self):
+        except AttributeError:
+            state = self.__dict__
+
         if type(self).__module__.startswith('sklearn.'):
```
This conditional is not properly covered, because the classes I use in the test are from within the sklearn module namespace (since scikit-learn has its tests in the sklearn namespace). So basically `return state.copy()` is never reached in the tests. Problem is, I cannot patch `type(self).__module__` without breaking the pickling mechanism (it makes a lookup for the module). So either we make the string `'sklearn.'` in the `BaseEstimator` mock-patchable, or we need to create a test-class outside of the sklearn namespace, to get this conditional fully covered.

HolgerPeters replied:
My idea was to write something like this

```
    def test_multiple_inheritance_setting_foreign_namespace(self):
        try:
            estimator = MultiInheritanceEstimator()
            old_mod = type(estimator).__module__
            type(estimator).__module__ = "notsklearn"

            serialized = pickle.dumps(estimator, protocol=2)
        finally:
            type(estimator).__module__ = old_mod
```
which doesn't work for the aforementioned reason.

jnothman replied:
I think we already test something like this. You can either test get/set_state directly (instead of pickling) or hack the pickle loading, perhaps by hacking `sys.modules`.

HolgerPeters replied:
+1 for testing __getstate__ / __setstate__ directly.

HolgerPeters commented:
Not quite sure why codecov reports a reduction in coverage. It seems all code paths are now fully tested.

<img width="643" alt="screen shot 2017-02-09 at 15 25 22" src="https://cloud.githubusercontent.com/assets/4097049/22787276/067add2a-eedc-11e6-9a29-c1e9c0b91429.png">

lesteve commented:
> Not quite sure why codecov reports a reduction in coverage

codecov seems to complain about a drop in coverage in `sklearn/test/test_base.py` looking at https://github.com/scikit-learn/scikit-learn/pull/8324#issuecomment-278593732.

Quickly looking at the [diff](https://codecov.io/gh/scikit-learn/scikit-learn/compare/3a0ea190c58394b080e84d200486c3f864822481...f08531439b855a1f25e8b4144fb0e4e97a7b5d73/src/sklearn/tests/test_base.py) it seems like you are not using `MultiInheritanceEstimator.cache` and `SingleInheritanceEstimator`.

HolgerPeters commented:
Alright, I think I have incorporated your feedback in this PR. And all CI was green. I have now squashed the commits, so it makes for a nicer patch in the history and rebased stuff on the most recent master. Would you say it is mergeable? Anything else I need to address?

jnothman commented:
You don't need to squash: github provides a "squash and merge" button. Also, unless there are merge conflicts, rebase is usually unnecessary (and even then, merging in the latest master suffices).

However, you generally require two full reviews and "LGTM"s before merge. We have a long backlog of reviewing. Thanks for your patience.

jnothman commented on sklearn/base.py:
```diff
@@ -290,10 +290,15 @@ def __repr__(self):
+            return dict(state.items(), _sklearn_version=__version__)
         else:
-            return dict(self.__dict__.items())
+            return state.copy()
```
Indeed, [an object's state need not be a dictionary](https://docs.python.org/2/library/pickle.html#object.__setstate__) and this line will break given some other types.

HolgerPeters replied:
Fixed in 13c74fc1284ec75e624c712f4ec23cf5fac5cacf

jnothman commented on sklearn/base.py:
```diff
@@ -305,7 +310,11 @@ def __setstate__(self, state):
+        try:
+            super(BaseEstimator, self).__setstate__(state)
+        except AttributeError:
+            self.__dict__.update(state)
```
I think this is okay, but FYI, pickle doesn't directly use update, in order to ensure all strings are interned: https://github.com/python/cpython/blob/master/Lib/pickle.py#L1522.

jnothman commented on sklearn/tests/test_base.py:
```diff
@@ -359,3 +357,81 @@ def test_pickle_version_warning():
+        return self._cache
+
+
+class TestPicklingConstraints(object):
```
1. We usually don't use test classes, but I'm personally okay with this as a way of grouping the code.
2. It might be good to mention that this test is about `BaseEstimator` somewhere (though I see there's a lack of that in this file).

HolgerPeters replied:
1. Addressed in 0a5541435b343454d79e44a7c3149908d82c43e7
2. since the tests are about estimators, isn't it (implicitly) clear, that `BaseEstimator` is involved. Also, the module being `sklearn.base`.

jnothman commented on sklearn/tests/test_base.py:
```diff
@@ -359,3 +357,81 @@ def test_pickle_version_warning():
+
+    def test_singleinheritance_clone(self):
+        estimator = SingleInheritanceEstimator()
+        assert estimator.cache
```
It's really hard to tell from the test code that this modifies `__dict__`. I think I'd rather something more transparent than a cache. For example, the getstate could store a timestamp.

jnothman commented on sklearn/tests/test_base.py:
```diff
@@ -359,3 +357,81 @@ def test_pickle_version_warning():
+
+        serialized = pickle.dumps(estimator, protocol=2)
+        estimator_restored = pickle.loads(serialized)
+        assert estimator_restored.b == 5
```
I notice this test tests the basic restoration that should have also been tested in `test_pickle_version_warning` which should really be checking that the loaded pickled trees can still predict. (You could add that if you wish.)

HolgerPeters replied:
Added in 393ff642714bfa723032abe016dd010fc7baf6bc

jnothman commented on sklearn/tests/test_base.py:
```diff
@@ -359,3 +357,81 @@ def test_pickle_version_warning():
+        finally:
+            type(estimator).__module__ = old_mod
+
+    def test_uses_object_dictionary_when_getstate_not_present(self):
```
I don't get what you mean by "when_getstate_not_present". You're using a `MultiInheritanceEstimator` for which `__getstate__` is present twice in the MRO.

HolgerPeters replied:
I am not sure what I meant by this test anymore, so I removed it in a94e4d1c51cb48950c56669a9117f69568989463 without decreasing coverage

HolgerPeters committed:
In BaseEstimator.__getstate__ copy only the dictionary

HolgerPeters committed:
Split up test_pickle_version_warning tests into several, independent test cases

HolgerPeters committed:
Add a check that the restored estimator can also make predictions

HolgerPeters committed:
Don't use test classes to group tests

HolgerPeters committed:
only use formatting for parts of the string that change

HolgerPeters committed:
Flake 8 column limit

jnothman commented on sklearn/tests/test_base.py:
```diff
@@ -314,48 +312,131 @@ def transform(self, X, y=None):
     tree = TreeBadVersion().fit(iris.data, iris.target)
     tree_pickle_other = pickle.dumps(tree)
     message = ("Trying to unpickle estimator TreeBadVersion from "
-               "version {0} when using version {1}. This might lead to "
```
I think it's clearer to the maintainer if "something" is still formatted in

HolgerPeters replied:
Hope a35832cb60cc08e2552251ef1d2c7ee6e50b4b6e is as you intend it to be. I really didn't like the `replace` calls in the old tests. I assume you favour a global message template, two over repeating it in the tests. The other option would be to duplicate the template in the tests (which I think is probably the least optimal variant).

jnothman commented on sklearn/tests/test_base.py:
```diff
@@ -314,48 +312,131 @@ def transform(self, X, y=None):
+        self._cache = None
+
+    @property
+    def cache(self):
```
I still don't think this makes for the test being easily read, and would rather something more explicit than a property with side effects. Even if the test did `estimator._cache = "some_value"` directly it would be better.

HolgerPeters replied:
Addressed in c965511e947e3c4618cb80e47e92e2fb5f83692a, indeed this is better than the property

jnothman reviewed (commented):
Thanks. In addition to these nitpicks, you've not tested your modifications to `__setstate__`

HolgerPeters committed:
Make the modification of the estimator more explicit in the tests

HolgerPeters committed:
As suggested in code review, prefer formatting over two literals

HolgerPeters committed:
Also assert, that __setstate__ overwriting works in mixin

lesteve commented on sklearn/tests/test_base.py:
```diff
@@ -314,48 +312,138 @@ def transform(self, X, y=None):
+        return data
+
+    @property
+    def cache(self):
```
You are still not using cache anywhere, right? If so can you please remove it?

HolgerPeters replied:
fixed in 32cbb3644a02021e30f6cb19b340f47077c5696e

lesteve commented on sklearn/tests/test_base.py:
```diff
@@ -314,48 +312,138 @@ def transform(self, X, y=None):
+
+    serialized = pickle.dumps(estimator, protocol=2)
+    estimator_restored = pickle.loads(serialized)
+    assert estimator_restored.b == 5
```
Using bare asserts with nose creates not so great error messages. This does matter on CIs. Can you please use `assert_*` helpers in `sklearn.utils.testing` in all your tests?

I know the migration to pytest is definitely on the radar, but still, I think this is the right thing to do. I'd be happy to hear different opinions.

An example of using bare asserts vs an `assert_*` helper when running with nose:
```
======================================================================
FAIL: test_nose.test
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/volatile/le243287/miniconda3/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/tmp/test_nose.py", line 7, in test
    assert x > y
AssertionError

======================================================================
FAIL: test_nose.test2
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/volatile/le243287/miniconda3/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/tmp/test_nose.py", line 13, in test2
    assert_greater(x, y)
AssertionError: 2 not greater than 4

----------------------------------------------------------------------
Ran 2 tests in 0.002s
```

HolgerPeters replied:
see e1417ad7a0bbe262efd3422cace9c79b85c4aefb

using py.test locally so I wasn't aware that this is an issue with nose :)

lesteve commented on sklearn/base.py:
```diff
@@ -290,10 +290,15 @@ def __repr__(self):
+
         if type(self).__module__.startswith('sklearn.'):
-            return dict(self.__dict__.items(), _sklearn_version=__version__)
+            return dict(state.items(), _sklearn_version=__version__)
```
Actually thinking about it I am a bit confused about this, should we not always have a warning mechanism even if the estimator is outside scikit-learn? I could imagine someone inheriting from say `LogisticRegression` with a minor modification and the warning applies to this case as well, right?

jnothman replied:
Firstly that is not a problem for this PR. Secondly, we do not issue the warning if out of sklearn because such an estimator is likely to be versioned differently.

lesteve replied:
> Firstly that is not a problem for this PR.

Agreed.

> Secondly, we do not issue the warning if out of sklearn because such an estimator is likely to be versioned differently.

OK I trust your judgement on this. The use case I had in mind was a thin wrapper around a scikit-learn estimator (deriving from an estimator class mostly for convenience), in which case it would make sense to get the warnings.

lesteve reviewed (commented):
A few comments

HolgerPeters committed:
Remove cache property

HolgerPeters committed:
Use assertion functions from sklearn.utils.testing

lesteve commented on sklearn/tests/test_base.py:
```diff
@@ -314,48 +313,132 @@ def transform(self, X, y=None):
+    estimator = MultiInheritanceEstimator()
+    estimator._cache = "this attribute should not be pickled"
+
+    serialized = pickle.dumps(estimator, protocol=2)
```
Why protocol=2? The tests pass fine with `serialized = pickle.dump(estimator)`.

HolgerPeters replied:
Removed in 7985b0b74b85839b1db2be85ed0ed624be6acc16

lesteve commented on sklearn/tests/test_base.py:
```diff
@@ -314,48 +313,132 @@ def transform(self, X, y=None):
+
+class MultiInheritanceEstimator(BaseEstimator, DontPickleCacheMixin):
+    def __init__(self, b=5):
+        self.b = b
```
I feel like the test could be made clearer by better variable naming, e.g. `attr_pickled` and `attr_not_pickled`. If you choose to do it, change it uniformly (and also maybe the naming of DontPickleCacheMixin).

lesteve commented on sklearn/tests/test_base.py:
```diff
@@ -314,48 +313,132 @@ def transform(self, X, y=None):
 
-    # check that not including any version also works:
+
+class TreeNoVersion(DecisionTreeClassifier):
```
My 2c: moving things around like this adds unnecessary noise in the diff thus making it harder to review without any significant benefit.

lesteve reviewed (commented):
Some small comments. LGTM otherwise.

HolgerPeters committed:
remove the protocol argument in tests

HolgerPeters committed:
Rename attributes to better convey their purpose

HolgerPeters committed:
Revert change of module in TreeNoVersion

HolgerPeters committed:
Adhere to column-limit

lesteve commented:
LGTM, could you add an entry in doc/whats_new.rst?

HolgerPeters committed:
changelog entry

lesteve commented on doc/whats_new.rst:
```diff
@@ -220,6 +220,10 @@ Bug fixes
      :issue:`7331` by `Joel Nothman`_ and :user:`yangarbiter`.
+   - Fix a bug where :meth:`sklearn.base.BaseEstimator.__getstate__` blocked
+     obstructed pickling customizations of child-classes, when used in a
+     multiple inheritence context.
```
typo: inherit**a**nce

HolgerPeters replied:
indeed, should be fixed with the last push.

lesteve commented on doc/whats_new.rst:
```diff
@@ -220,6 +220,10 @@ Bug fixes
    - Fix a bug in cases where `numpy.cumsum` may be numerically unstable,
      raising an exception if instability is identified.  :issue:`7376` and
      :issue:`7331` by `Joel Nothman`_ and :user:`yangarbiter`.
+   - Fix a bug where :meth:`sklearn.base.BaseEstimator.__getstate__` blocked
```
looks like you could not make up your mind between blocked and obstructed ;-)

lesteve reviewed (approved):
LGTM, @jnothman do you want to have a look at this one so we can merge it? Two minor comments in the changelog.

HolgerPeters committed:
Fix commit message

jnothman commented:
LGTM thanks @holgerpeters

lesteve commented:
Great stuff @HolgerPeters, thanks a lot!
